opendevin/server/session/agent.py

```diff
@@ -0,0 +1,110 @@
+from typing import Optional
+
+from agenthub.codeact_agent.codeact_agent import CodeActAgent
+from opendevin.controller import AgentController
+from opendevin.controller.agent import Agent
+from opendevin.controller.state.state import State
+from opendevin.core.config import config
+from opendevin.core.logger import opendevin_logger as logger
+from opendevin.core.schema import ConfigType
+from opendevin.events.stream import EventStream
+from opendevin.llm.llm import LLM
+from opendevin.runtime import DockerSSHBox
+from opendevin.runtime.e2b.runtime import E2BRuntime
+from opendevin.runtime.runtime import Runtime
+from opendevin.runtime.server.runtime import ServerRuntime
+
+
+class AgentSession:
+    """Represents a session with an agent.
+
+    Attributes:
+        controller: The AgentController instance for controlling the agent.
+    """
+
+    sid: str
+    event_stream: EventStream
+    controller: Optional[AgentController] = None
+    runtime: Optional[Runtime] = None
+
+    def __init__(self, sid):
+        """Initializes a new instance of the Session class."""
+        self.sid = sid
+        self.event_stream = EventStream(sid)
+
+    async def start(self, start_event: dict):
+        """Starts the agent session.
+
+        Args:
+            start_event: The start event data (optional).
+        """
+        if self.controller or self.runtime:
+            raise Exception(
+                'Session already started. You need to close this session and start a new one.'
+            )
+        await self._create_runtime()
+        await self._create_controller(start_event)
+
+    async def close(self):
+        if self.controller is not None:
+            end_state = self.controller.get_state()
+            end_state.save_to_session(self.sid)
+            await self.controller.close()
+        if self.runtime is not None:
+            self.runtime.close()
+
+    async def _create_runtime(self):
+        if self.runtime is not None:
+            raise Exception('Runtime already created')
+        if config.runtime == 'server':
+            logger.info('Using server runtime')
+            self.runtime = ServerRuntime(self.event_stream, self.sid)
+        elif config.runtime == 'e2b':
+            logger.info('Using E2B runtime')
+            self.runtime = E2BRuntime(self.event_stream, self.sid)
+
+    async def _create_controller(self, start_event: dict):
+        """Creates an AgentController instance.
+
+        Args:
+            start_event: The start event data (optional).
+        """
+        if self.controller is not None:
+            raise Exception('Controller already created')
```

Comment:
In the old `server/agent/agent.py`, we have the following code:

```python
        if self.controller is not None:
            await self.controller.close()
```

I don't understand why we needed that at the first place, but I also wonder why you changed the logic here.

Reply:
In the old version, you were allowed to reuse a session with new agent controllers. In the new one, if you want to reinitialize the agent, you need a new session.

Reply:
good to know, thanks!